### PR TITLE
[Audio metrics 03/12] Add metric tokenization and constrained ARECHO search

### PR DESCRIPTION
**Pull request by ftshijt**
Adds metric tokenization and constrained ARECHO decoding while preserving published vocabulary IDs, numeric bins, and cumulative search scores.

Part 3/12 of the Uni-VERSA / ARECHO revival; prerequisite #6749 has merged. The branch is aligned with upstream `master` at `a4851823e8b` (PR merge commit `05bb33e51ef`). The series index is #6747. The PR contains metric tokenization/search plus small model-contract fixes requested in review (ten files).

Token decoding validates metric/value ownership and token bounds, consistently handles the numeric zero bin, and accepts tensor, list, and iterator sequences. Restoring metric-local IDs includes the four special-token IDs. Constrained search preserves decoder state even when meta-label scores are skipped, and supports fixed or free metric order and oversized beams.

The beam search is a registered PyTorch module for callable/device/dtype support, rejects invalid metric IDs and mask ranges, and the abstract tokenizer signatures match the concrete API. CI docstring H404/H405 failures are fixed.

The simplification removes the duplicate metric dictionary, duplicate sequence-value decoder, deep-copy helper for integer labels, and repeated beam-expansion setup. Historical special-token IDs, numeric threshold behavior, and unrestricted-search fallback remain intact.

Validation:
- 110 PR-local Uni-VERSA and metric-data tests pass after merging upstream master.
- All 34 focused tokenizer/search tests also pass on Python 3.12 / CPU PyTorch 2.11.
- 120 tests pass against an isolated checkout of the complete series, including downstream ARECHO training/inference, metric data handling, and the inference CLI.
- Repository-wide pre-commit, pycodestyle, and both CI flake8 invocations (with hacking and flake8-docstrings) pass; the CI image configuration invariant checker and git diff --check pass.
- Broader local model tests use Python 3.10 / CPU PyTorch 2.11; focused tests also pass on Python 3.12. All 76 hosted checks pass on the review follow-up `f74f573e1b9`: Ubuntu unit/configuration/integration suites, Debian, macOS, Windows, documentation, pre-commit, and Codecov. No failed or pending checks remain.

Review follow-up (`f74f573e1b9`): `AbsUniversa` now declares default reference-use flags and optional reference arguments; normalization preserves caller-owned audio/reference tensors; missing reference lengths raise clear `ValueError`s. Seven new regressions pass on Python 3.12; six failed against the previous code. Four implemented inline findings are marked resolved.

Task registration is intentionally tracked by [PR #6752](https://github.com/espnet/espnet/pull/6752), the series' task component. Its `AudioMetricTask` registers `UniversaBase`/ARECHO through `ClassChoices` and constructs `ESPnetUniversaModel`; its task construction/legacy alias tests pass in the downstream validation. That review thread remains open here until its disposition is accepted, rather than duplicating the later component into this PR.

### CODE REVIEW

File: espnet2/universa/abs_universa.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        self,
+        audio: torch.Tensor,
+        audio_lengths: torch.Tensor,
+        metrics: Dict[str, torch.Tensor],
+        ref_audio: torch.Tensor,
+        ref_audio_lengths: torch.Tensor,
+        ref_text: torch.Tensor,
+        ref_text_lengths: torch.Tensor,
+        **kwargs,
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor], torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/abs_universa.py:30]: Fixed in [f74f573](https://github.com/ftshijt/espnet/commit/f74f573e1b9ad7960f53e1960bd34159eb26a62c). `AbsUniversa` now declares `use_ref_audio = False` and `use_ref_text = False`, and all four reference arguments in its abstract `forward` signature are optional with `None` defaults. Added a regression that constructs a reference-free subclass through `ESPnetUniversaModel` and exercises both training and inference without reference attributes or inputs. The regression failed before the fix and now passes; the full hosted CI matrix is green.

File: espnet2/universa/ar_universa/universa_beam_search.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+    """
+
+    yseq: torch.Tensor
+    score: Union[float, torch.Tensor] = 0
+    scores: Dict[str, Union[float, torch.Tensor]] = dict()
+    states: Dict[str, Any] = dict()
+    unused_meta_label_ids: List[int] = None
+
+
+class ARUniVERSABeamSearch:
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/ar_universa/universa_beam_search.py]: Fixed in [6566bcb](https://github.com/ftshijt/espnet/commit/6566bcbd060e8711e1dc23cba9a5db2012880431). `ARUniVERSABeamSearch` now inherits from `torch.nn.Module`, so its scorer `ModuleDict` participates in recursive conversion and `search(x)` works. The new regression checks a registered scorer buffer moves to float64 and verifies callable decoding. Downstream ARECHO training/inference tests and the full hosted CI matrix pass.

File: espnet2/universa/ar_universa/universa_beam_search.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        # set scorers
+        if beam_size < 1:
+            raise ValueError("beam_size must be positive")
+        if len(set(meta_label_for_search)) != len(meta_label_for_search):
+            raise ValueError("Requested metrics must be unique")
+        self.use_fixed_order = use_fixed_order
+        self.weights = weights
+        self.meta_label_for_search = meta_label_for_search
+        self.skip_meta_label_score = skip_meta_label_score
+        self.beam_masking = beam_masking
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/ar_universa/universa_beam_search.py]: Fixed in [6566bcb](https://github.com/ftshijt/espnet/commit/6566bcbd060e8711e1dc23cba9a5db2012880431). Construction now rejects metric labels outside `[0, vocab_size)` and masking intervals unless `0 <= start < end <= vocab_size`. Parameterized regressions cover negative/out-of-range labels and negative, empty, reversed, and out-of-range intervals; all fail before the fix and pass afterward. Existing constrained-search/order/score tests and full CI remain green.

File: espnet2/universa/espnet_model.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        self,
+        universa: AbsUniversa,
+        frontend: AbsFrontend,
+    ):
+        """Initialize ESPnet model for Universa."""
+        super().__init__()
+        self.frontend = frontend
+        self.universa = universa
+        self.use_ref_audio = universa.use_ref_audio
+        self.use_ref_text = universa.use_ref_text
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/espnet_model.py:30]: This registration is implemented in the separately scoped series component [PR #6752](https://github.com/espnet/espnet/pull/6752), rather than duplicated into this tokenizer/search PR. Its `AudioMetricTask` registers both `UniversaBase` and ARECHO through `ClassChoices`, constructs `ESPnetUniversaModel`, and retains the historical `UniversaTask` alias. I verified the task source matches #6752 and ran its construction/portable-vocabulary/legacy-alias test as part of the 120 passing downstream-series tests. #6752 is still open, so this capability is not claimed to be merged here. Leaving this thread open to make that cross-PR disposition visible.

File: espnet2/universa/metric_tokenizer/metric_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+class AbsMetricTokenizer(ABC):
+    @abstractmethod
+    def metric2token(
+        self, metrics: Dict[str, Union[float, str, int]]
+    ) -> List[Tuple[int, int]]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def token2metric(self, tokens: Iterable[int]) -> str:
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/metric_tokenizer/metric_tokenizer.py]: Fixed in [6566bcb](https://github.com/ftshijt/espnet/commit/6566bcbd060e8711e1dc23cba9a5db2012880431). `AbsMetricTokenizer.metric2token` now matches the concrete dictionary return type, accepted token pairs, and `reduce_offset` argument; `token2metric` takes a single token and optional metric; `tokenseq2metric` is a separate abstract method with the concrete sequence-decoding signature. Tokenizer/search regressions pass on Python 3.12 and the full hosted CI matrix is green.

**Comment by ftshijt** [comment]
Review follow-up is complete for the fixes implemented in this PR, through [f74f573](https://github.com/ftshijt/espnet/commit/f74f573e1b9ad7960f53e1960bd34159eb26a62c). I have replied to each of the five inline threads with the relevant commit, behavior, and validation.

The two additional findings in the review summary are also fixed in f74f573:
- **Caller-owned audio mutation:** normalization now receives cloned audio and reference tensors. Regressions verify both inputs remain unchanged under no-grad inference and that gradients still reach the original inputs during training.
- **Missing reference lengths:** the wrapper now raises explicit `ValueError`s naming `ref_audio_lengths` or `ref_text_lengths`. All four audio/text and training/inference combinations are covered.

**Validation on f74f573:** all 76 hosted checks pass, including the complete Ubuntu PR matrix, Debian, macOS, Windows, docs, pre-commit, and Codecov. Patch coverage is **89.68%**, above the **73.16%** target; project coverage is **73.24%**. The Codecov comment's missing-line notice does not mean the patch check failed. Local validation passes 110 component/data tests, 120 downstream-series tests, and all seven new review regressions on Python 3.12.

Four implemented inline findings are resolved. **Task registration remains tracked in #6752**, where its implementation and tests already exist; that PR is still open and its thread here remains open to avoid presenting deferred work as merged.

The earlier CodeRabbit walkthrough, risk assessment, and docstring-coverage warning were generated against da10c31. The CI H404/H405 docstring violations have been fixed; a fresh review should reassess the current diff and its separate docstring-coverage metric.

@coderabbitai full review

**Comment by sw005320** [comment]
One architectural question: could this constrained search be implemented on top of the existing ESPnet `BeamSearch`/scorer abstraction rather than introducing a new `ARUniVERSABeamSearch`?

Conceptually, the metric-aware constraint could be implemented as an additional scorer or token mask:

```text
constraint_score(token) =
    0       if token is allowed
    -inf    otherwise
```

where the allowed token set depends on the current metric, metric order, and decoding state.

I understand that this implementation also needs to maintain metric-specific state and optionally skip meta-label scores, so there may be a good reason that the existing `BeamSearch` interface is not expressive enough.

If so, could we briefly document what limitation of the existing BeamSearch/scorer abstraction requires a separate search implementation?

I would prefer reusing the existing beam-search machinery if the constraint can be expressed cleanly as a scorer/stateful constraint, since otherwise we may end up maintaining two very similar beam-search implementations.

**Comment by sw005320** [comment]
Other comments (questions)

1. **Mutable defaults in `Hypothesis`**
   `scores` and `states` currently use `dict()` as default values. Since these are mutable objects, different `Hypothesis` instances can potentially share the same dictionaries. Could we use per-instance defaults here? This is probably worth fixing since hypotheses are created extensively during beam search.

2. **EOS / search termination**
   `ARUniVERSABeamSearch` keeps an `eos` argument, but the actual search terminates based on `len(meta_label_for_search)` rather than EOS. Is EOS intentionally unused because the output structure is always a fixed number of metric/value pairs? If so, it might be worth documenting this explicitly or simplifying the interface.

3. **Unknown metrics in `metric2token()`**
   `metric2token()` silently skips metrics that are not present in `tokenizer_config`. I understand that `tokenize_metric` may intentionally select a subset of metrics, but could we distinguish an intentionally unselected metric from an unknown/invalid metric name? Silently dropping a misspelled metric could make data/configuration errors difficult to detect.

**Comment by sw005320** [comment]
> One architectural question: could this constrained search be implemented on top of the existing ESPnet `BeamSearch`/scorer abstraction rather than introducing a new `ARUniVERSABeamSearch`?
> 
> Conceptually, the metric-aware constraint could be implemented as an additional scorer or token mask:
> 
> ```
> constraint_score(token) =
>     0       if token is allowed
>     -inf    otherwise
> ```
> 
> where the allowed token set depends on the current metric, metric order, and decoding state.
> 
> I understand that this implementation also needs to maintain metric-specific state and optionally skip meta-label scores, so there may be a good reason that the existing `BeamSearch` interface is not expressive enough.
> 
> If so, could we briefly document what limitation of the existing BeamSearch/scorer abstraction requires a separate search implementation?
> 
> I would prefer reusing the existing beam-search machinery if the constraint can be expressed cleanly as a scorer/stateful constraint, since otherwise we may end up maintaining two very similar beam-search implementations.

One benefit of using existing beam-search is that it is well maintained and also supports fast batch beam search (across both utterances and hypotheses). This would be a strong point.

**Comment by ftshijt** [comment]
@sw005320 Thank you for the architectural feedback and follow-up questions. Addressed in [4148a8f7f55](https://github.com/ftshijt/espnet/commit/4148a8f7f55):

1. **Shared search machinery:** `ARUniVERSABeamSearch` now subclasses `BeamSearch` and calls its `search()` for both token phases. A `MetricConstraintScorer` implements the proposed zero/negative-infinity mask. Scorer registration, hypothesis construction, token expansion, and score/state merging all come from the shared implementation. This removes 206 net production lines.
2. **Pair scheduling and batching:** the remaining wrapper preserves global pruning *after a complete metric/value pair*. In particular, skipping label scores must retain every allowed label until its value is scored, even with beam size 1. A normal token-level beam can discard the winning pair early; a regression covers that case. The class documents this and fixed-pair termination as the two reasons for the wrapper. You are right that `BatchBeamSearch` supports both utterances and hypotheses. The constraint implements its batch scorer interface and passes a test with two utterances and two hypotheses each. The pair wrapper still takes one utterance; using the batch engine for that complete wrapper also requires adapting its pruning/termination schedule. This change does not claim batched ARECHO inference.
3. **Mutable defaults and EOS:** the duplicate local `Hypothesis` and its shared dictionary defaults are removed. The shared engine explicitly supplies fresh score/state dictionaries; regression coverage checks independent initial hypotheses and siblings. Unused-metric lists are also removed: the emitted prefix is the source of truth. EOS is retained for decoder-configuration compatibility and explicitly documented as neither appended nor scored; output is SOS plus the requested number of pairs.
4. **Unknown metric names:** unknown input names and unknown entries in `tokenize_metric` raise `ValueError` naming the metric. Known metrics omitted from the selected subset are still intentionally skipped, with tests distinguishing these cases.

Local validation: **118** component/data tests, **128** full-series downstream tests (including ARECHO training/inference and the CLI), **42** focused tests on Python 3.12, **240** deterministic comparisons against the original search, repository-wide pre-commit/style/CI lint, and the CI configuration invariant checker all pass. The full hosted CI matrix is now running for this new commit; the prior head had all 76 checks passing. I will report the new head's results separately.

@coderabbitai full review

**Comment by ftshijt** [comment]
CI follow-up: the new batch-scorer integration test exposed a PyTorch-dependent tie choice in the shared batch engine: an impossible padded beam can contribute an early EOS to its returned list. The test had set only the maximum output length. [6c25b044aba](https://github.com/ftshijt/espnet/commit/6c25b044aba) also sets the minimum length required by the fixed-pair protocol, retaining the exact hypothesis-count, sequence, and score assertions. No production code changed in this follow-up.

I reproduced the failing padded-EOS selection deterministically locally: the old test configuration returns three hypotheses, while the corrected one returns the two expected complete pairs. All 42 focused tests and relevant pre-commit hooks pass. The full hosted matrix is rerunning on the corrected head. CodeRabbit's full review of the production refactor reported no new correctness, reproducibility, or API-compatibility problems.

**Comment by ftshijt** [comment]
Validation is complete for **6c25b044aba**: **all 76 checks pass**, with no failed or pending checks. The merge-queue check is neutral and maintainer review is still required.

- [Ubuntu](https://github.com/espnet/espnet/actions/runs/35725891298): full unit/configuration/integration matrix passes, including all six ESPnet2 Python/PyTorch combinations and the previously failing Python 3.12 / PyTorch 2.13 job.
- [Debian](https://github.com/espnet/espnet/actions/runs/35725891247), [macOS](https://github.com/espnet/espnet/actions/runs/35725891211), and [Windows](https://github.com/espnet/espnet/actions/runs/35725891271): all aggregate gates pass.
- Documentation, pre-commit, and both Codecov checks pass. Patch coverage is **88.89%**; project coverage is **73.23%**. The earlier red coverage status was from incomplete uploads.
- Local validation: **118** component/data tests, **128** downstream-series tests, **42** focused Python 3.12 tests, and **240** comparisons against the original search pass.

The architectural and API feedback is addressed as detailed above; CodeRabbit's fresh full review of the production refactor found no new issues. The complete pair decoder retains its single-utterance API, with the batching/scheduling limitation explicitly documented. Task registration remains tracked in #6752, and that scope-disposition thread remains open here.

**Comment by sw005320** [comment]
This looks much cleaner now. I especially like the separation between metric tokenization, constrained search, the Uni-VERSA model, and the task registration in #6752. The reference-input handling and input-ownership issues also look properly addressed.

I only noticed one small issue in the current head: could we avoid mutable default dictionaries in `Hypothesis` for `scores` and `states`? If these are defined as `{}` at the class/dataclass level, multiple hypotheses constructed without explicit values can share the same dictionaries. Using `field(default_factory=dict)` would preserve the current API while making each hypothesis independent.

Other than that, I think this is in good shape. The current test/CI coverage also looks sufficient to me.

**Comment by ftshijt** [comment]
@sw005320 You’re right: the refactor removed the local duplicate, but the shared `Hypothesis` constructor still had the same mutable-default issue for callers omitting these fields. Fixed at the shared source in [f6f40573d0c](https://github.com/ftshijt/espnet/commit/f6f40573d0c).

`scores` and `states` now receive a fresh dictionary per construction; `hs` also receives a fresh list. Explicitly supplied containers retain their identity. I kept the named-tuple representation because existing scalar/batch search and device-transfer code use tuple operations, `_replace`, and `_asdict`. A per-construction factory in `__new__` provides the same isolation as `default_factory` without changing that API.

Validation: five new regressions cover default isolation, explicit-container ownership, tuple replacement, JSON conversion, pickle round trips, and device transfer. They pass on Python 3.12. The existing scalar/batch/streaming search suites pass **313 tests** (198 existing GPU/unsupported-configuration skips); **123** hypothesis/component/data tests and **128** downstream-series tests also pass. Both repository-wide CI lint commands and the affected pre-commit hooks pass.

Full hosted CI is running for this commit. The prior head had all 76 checks passing; I’ll update this with the new head’s results.

File: espnet2/universa/espnet_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ftshijt) [reply espnet2/universa/espnet_model.py:30]: Shinji has now explicitly endorsed the separation of task registration into #6752 in [his latest review comment](https://github.com/espnet/espnet/pull/6750#issuecomment-5777113344). Resolving this thread as an accepted scope clarification. Task registration remains owned by #6752; this does not claim that registration is merged or provided by this tokenizer/search PR.

**Review by sw005320 — APPROVED** [review]
thanks!
